contracts/oracle/CollateralizationMemoizer.sol

```diff
@@ -0,0 +1,204 @@
+                bool _validityStatus
+            ) = ICollateralizationOracle(collateralizationOracle).pcvStats();
+
+            require(_validityStatus, "CollateralizationMemoizer: CollateralizationOracle reading is invalid");
+
+            Decimal.D256 memory _thresholdLow = Decimal.from(10_000 - deviationThreshold).div(10_000);
+            Decimal.D256 memory _thresholdHigh = Decimal.from(10_000 + deviationThreshold).div(10_000);
+
+            // PCV < threshold
+            Decimal.D256 memory pcvRatio = Decimal.from(_protocolControlledValue).div(cachedProtocolControlledValue);
+            outdated = outdated || pcvRatio.lessThan(_thresholdLow);
+            // PCV > threshold
+            if (!outdated) {
+                outdated = outdated || pcvRatio.greaterThan(_thresholdHigh);
+            }
+            // CircFEI < threshold
+            if (!outdated) {
+                Decimal.D256 memory circFeiRatio = Decimal.from(_userCirculatingFei).div(cachedUserCirculatingFei);
+                outdated = outdated || circFeiRatio.lessThan(_thresholdLow);
+                // CircFEI > threshold
+                if (!outdated) {
+                    outdated = outdated || circFeiRatio.greaterThan(_thresholdHigh);
+                }
+            }
+            // no need to check protocol equity, it is deduced from the other 2
+        }
+    }
+
+    /// @notice Get the current collateralization ratio of the protocol, from cache.
+    /// @return collateralRatio the current collateral ratio of the protocol.
+    /// @return validityStatus the current oracle validity status (false if any
+    ///         of the oracles for tokens held in the PCV are invalid, or if
+    ///         this contract is paused).
+    function read() public view override returns (Decimal.D256 memory collateralRatio, bool validityStatus) {
+        collateralRatio = Decimal.ratio(cachedProtocolControlledValue, cachedUserCirculatingFei);
+        validityStatus = block.timestamp <= lastUpdate + validityDuration;
+    }
+
+    // ----------- ICollateralizationOracle override methods -----------
+
+    /// @notice returns the Protocol-Controlled Value, User-circulating FEI, and
+    ///         Protocol Equity. If there is a fresh cached value, return it.
+    ///         Otherwise, call the CollateralizationOrache to get fresh data.
+    /// @return protocolControlledValue : the total USD value of all assets held
+    ///         by the protocol.
+    /// @return userCirculatingFei : the number of FEI not owned by the protocol.
+    /// @return protocolEquity : the difference between PCV and user circulating FEI.
+    ///         If there are more circulating FEI than $ in the PCV, equity is 0.
+    /// @return validityStatus : the current oracle validity status (false if any
+    ///         of the oracles for tokens held in the PCV are invalid, or if
+    ///         this contract is paused).
+    function pcvStats() external returns (
```

Comment:
This should be a view method, likely split in two. pcvStats should simply return the cached values with a validity status equal to whether it is out of date or not

A second view method maybe called pcvStatsCurrent passes through to the CR oracle and returns most up to date values